ada_feeding_perception/ada_feeding_perception/table_detection.py

```diff
@@ -0,0 +1,483 @@
+        self.latest_depth_img_msg = None
+        self.latest_depth_img_msg_lock = threading.Lock()
+        aligned_depth_topic = "~/aligned_depth"
+        try:
+            aligned_depth_type = get_img_msg_type(aligned_depth_topic, self)
+        except ValueError as err:
+            self.get_logger().warn(
+                f"Error getting type of depth image topic. Defaulting to Image. {err}"
+            )
+            aligned_depth_type = Image
+        self.depth_image_subscriber = self.create_subscription(
+            aligned_depth_type,
+            aligned_depth_topic,
+            self.depth_image_callback,
+            1,
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+        # Subscribe to the RGB image topic, to store the latest image
+        self.latest_img_msg = None
+        self.latest_img_msg_lock = threading.Lock()
+        image_topic = "~/image"
+        try:
+            image_type = get_img_msg_type(image_topic, self)
+        except ValueError as err:
+            self.get_logger().warn(
+                f"Error getting type of image topic. Defaulting to CompressedImage. {err}"
+            )
+            image_type = CompressedImage
+        self.image_subscriber = self.create_subscription(
+            image_type,
+            image_topic,
+            self.image_callback,
+            1,
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+    def fit_table(
+        self,
+        image_msg: Image,
+        image_depth_msg: Image,
+    ):
+        """
+        Fits a plane to the table based on the given RGB image and depth image messages.
+
+        Parameters
+        ----------
+        image_msg: The RGB image to detect plates from.
+        image_depth_msg: The depth image corresponding to the RGB image.
+
+        Returns
+        ----------
+        table: The depth array of the table.
+        """
+        # Convert ROS images to CV images
+        image = ros_msg_to_cv2_image(image_msg, self.bridge)
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+
+        image_depth = ros_msg_to_cv2_image(image_depth_msg, self.bridge)
+
+        # Detect Largest Circle (Plate)
+        circles = cv2.HoughCircles(
+            gray,
+            cv2.HOUGH_GRADIENT,
+            self._hough_accum,
+            self._hough_min_dist,
+            param1=self._hough_param1,
+            param2=self._hough_param2,
+            minRadius=self._hough_min,
+            maxRadius=self._hough_max,
+        )
+        if circles is None:
+            return None, None, None
+        circles = np.round(circles[0, :]).astype("int")
+        plate_uv = (0, 0)
+        plate_r = 0
+        print(len(circles))
+        for x, y, r in circles:
+            print("Radius: " + str(r))
+            if r > plate_r:
+                plate_uv = (x, y)
+                plate_r = r
+
+        # Testing - Draw out the circles
```

Comment:
This is never used in your current code, right? In that case, it is a waste of computational resources. I'd suggest adding a `viz` boolean ROS2 parameter, and only if it is set to `true` do you do this. 

Also, in general you shouldn't rely on people (un)commenting code to get debugging capabilities. So when `viz` is true, you should instead publish the image to a ROS topic, which the person can then view in a tool like RVIZ. See face_detection for an example.